### PR TITLE
Sniper

### PR DESCRIPTION
**Pull request by Wouterz90**
I think the implemented way of using abilities with charges is the easiest to maintain, and to implement for new users. Would like to hear your opinions on this.

### CODE REVIEW

File: game/scripts/npc/npc_abilities_custom.txt [hunk 1]
```diff
@@ -2,8 +2,13 @@
...
 #base "heroes/skywrath_mage.kv"
-
+#base "heroes/sniper.kv"
 "DOTAAbilities"
 {
     "Version"   "1"
+    "base_ability_charges"
+    {
+        "BaseClass"                 "ability_lua"
+        "ScriptFile"            "heroes/util/base_ability_charges.lua"
```
Comment (Perryvw) [inline game/scripts/npc/npc_abilities_custom.txt]: this indentation, ocd = triggered

File: game/scripts/npc/npc_abilities_custom.txt [hunk 2]
```diff
@@ -2,8 +2,13 @@
 #base "heroes/antimage.kv"
 #base "heroes/tusk.kv"
 #base "heroes/skywrath_mage.kv"
-
+#base "heroes/sniper.kv"
```
Comment (Perryvw) [inline game/scripts/npc/npc_abilities_custom.txt]: keep open line here

File: game/scripts/vscripts/heroes/sniper/assassinate.lua [hunk 3]
```diff
@@ -0,0 +1,206 @@
+LinkLuaModifier("modifier_sniper_assassinate_caster_lua","heroes/sniper/assassinate.lua",LUA_MODIFIER_MOTION_NONE)
+LinkLuaModifier("modifier_sniper_assassinate_target_lua","heroes/sniper/assassinate.lua",LUA_MODIFIER_MOTION_NONE)
+
+-- The spell is almost completely different with and without Aghanims Scepter. The order is always normal then scepter
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/sniper/assassinate.lua]: Always? Where else?

File: game/scripts/vscripts/heroes/sniper/assassinate.lua [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+function sniper_assassinate_lua:GetBehavior()
+    if not self:GetCaster():HasScepter() then
+        return DOTA_ABILITY_BEHAVIOR_UNIT_TARGET
+    else
+        return DOTA_ABILITY_BEHAVIOR_POINT + DOTA_ABILITY_BEHAVIOR_AOE
+    end
+end
+---@override
+function sniper_assassinate_lua:GetAbilityDamageType()
+    if not self:GetCaster():HasScepter() then
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/sniper/assassinate.lua]: get rid of the not here and swap cases

File: game/scripts/vscripts/heroes/sniper/assassinate.lua [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+---@override
+function sniper_assassinate_lua:OnAbilityPhaseInterrupted()
+    -- Remove the crosshairs from the target(s), and remove the modifier from the caster
+    if self.storedTarget then
+        for k,v in pairs(self.storedTarget) do
+            v:RemoveModifierByName("modifier_sniper_assassinate_target_lua")
+        end
+    end
+    self.storedTarget = nil
+    self:GetCaster():RemoveModifierByName("modifier_sniper_assassinate_caster_lua")
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/sniper/assassinate.lua]: Should probably remove by name and caster.

File: game/scripts/vscripts/heroes/sniper/shrapnel.lua [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+---@class modifier_sniper_shrapnel_lua_aura : CDOTA_Modifier_Lua
+modifier_sniper_shrapnel_lua_aura = class({})
+---@override
+function modifier_sniper_shrapnel_lua_aura:OnCreated()
+    if IsServer() then
+        -- Ability specials
+        self.radius = self:GetAbility():GetSpecialValueFor("radius")
+
+        -- Add shrapnel particl effect
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/sniper/shrapnel.lua]: typo

File: game/scripts/vscripts/heroes/sniper/shrapnel.lua [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    end
+end
+
+function modifier_sniper_shrapnel_lua_aura_modifier:OnIntervalThink()
+    local damageTable = {
+        victim = self:GetParent(),
+        attacker = self:GetCaster(),
+        damage = self.damage,
+        damage_type = DAMAGE_TYPE_MAGICAL,
+    }
```
Comment (Perryvw) [inline game/scripts/vscripts/heroes/sniper/shrapnel.lua:108]: `ApplyDamage(damageTable)`?

**Review by Perryvw — CHANGES_REQUESTED** [review]
Needs some small changes, looks good overall

File: game/scripts/vscripts/heroes/sniper/shrapnel.lua (hunk 7 above)
Comment (Wouterz90) [reply game/scripts/vscripts/heroes/sniper/shrapnel.lua:108]: Not sure why this still shows

File: game/scripts/npc/heroes/sniper.kv [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+"DOTAAbilities"
+{
+	"sniper_shrapnel_lua"
+	{
+		// General
+		//-------------------------------------------------------------------------------------------------------------
+		"BaseClass"                     "ability_lua"
+	    "ScriptFile"                    "heroes/sniper/shrapnel.lua"
+	    "AbilityTextureName"      "sniper_shrapnel"
+		"AbilityBehavior"       "DOTA_ABILITY_BEHAVIOR_AOE | DOTA_ABILITY_BEHAVIOR_POINT | DOTA_ABILITY_BEHAVIOR_IGNORE_BACKSWING"
```
Comment (Perryvw) [inline game/scripts/npc/heroes/sniper.kv]: indentation

File: game/scripts/npc/heroes/sniper.kv [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+	// Sniper: Headshot
+	//=================================================================================================================
+	"sniper_headshot_lua"
+	{
+		// General
+		//-------------------------------------------------------------------------------------------------------------
+		"BaseClass"                     "ability_lua"
+	    "ScriptFile"                    "heroes/sniper/headshot.lua"
+	    "AbilityTextureName"      "sniper_headshot"
+		"AbilityBehavior"       "DOTA_ABILITY_BEHAVIOR_PASSIVE"
```
Comment (Perryvw) [inline game/scripts/npc/heroes/sniper.kv]: indentation

File: game/scripts/npc/heroes/sniper.kv [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+	//=================================================================================================================
+	// Sniper: Take Aim
+	//=================================================================================================================
+	"sniper_take_aim_lua"
+	{
+		// General
+		//-------------------------------------------------------------------------------------------------------------
+		"BaseClass"                     "ability_lua"
+	    "ScriptFile"                    "heroes/sniper/take_aim.lua"
+	    "AbilityTextureName"      "sniper_take_aim"
```
Comment (Perryvw) [inline game/scripts/npc/heroes/sniper.kv]: indentation

File: game/scripts/npc/heroes/sniper.kv [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+	//=================================================================================================================
+	// Sniper: Assassinate
+	//=================================================================================================================
+	"sniper_assassinate_lua"
+	{
+		// General
+		//-------------------------------------------------------------------------------------------------------------
+		"BaseClass"                     "ability_lua"
+	    "ScriptFile"                    "heroes/sniper/assassinate.lua"
+	    "AbilityTextureName"      "sniper_assassinate"
```
Comment (Perryvw) [inline game/scripts/npc/heroes/sniper.kv]: indentation

File: game/scripts/npc/npc_abilities_custom.txt [hunk 12]
```diff
@@ -3,7 +3,13 @@
 #base "heroes/tusk.kv"
 #base "heroes/skywrath_mage.kv"
 
+#base "heroes/sniper.kv"
```
Comment (Perryvw) [inline game/scripts/npc/npc_abilities_custom.txt]: I meant keep an open line between the #base lines and DOTAAbilities, so move this to line 5 and keep 6 open.

**Review by Perryvw — CHANGES_REQUESTED** [review]
some formatting issues

**Review by Perryvw — APPROVED** [review]

**Pull request merged**